Sources/PackageLoading/ManifestLoader.swift

```diff
@@ -1079,6 +1034,110 @@ public final class ManifestLoader: ManifestLoaderProtocol {
+        ) throws {
+            let manifestContents = try fileSystem.readFileContents(manifestPath).contents
+            let sha256Checksum = try Self.computeSHA256Checksum(packageIdentity: packageIdentity, manifestContents: manifestContents, toolsVersion: toolsVersion, env: env, swiftpmVersion: swiftpmVersion)
+
+            self.packageIdentity = packageIdentity
+            self.manifestPath = manifestPath
+            self.manifestContents = manifestContents
+            self.toolsVersion = toolsVersion
+            self.env = env
+            self.swiftpmVersion = swiftpmVersion
+            self.sha256Checksum = sha256Checksum
+        }
+
+        func hash(into hasher: inout Hasher) {
+            hasher.combine(self.sha256Checksum)
+        }
+
+        private static func computeSHA256Checksum(
+            packageIdentity: PackageIdentity,
+            manifestContents: [UInt8],
+            toolsVersion: ToolsVersion,
+            env: EnvironmentVariables,
+            swiftpmVersion: String
+        ) throws -> String {
+            let stream = BufferedOutputByteStream()
+            stream <<< packageIdentity
+            stream <<< manifestContents
+            stream <<< toolsVersion.description
+            for (key, value) in env.sorted(by: { $0.key > $1.key }) {
+                stream <<< key <<< value
+            }
+            stream <<< swiftpmVersion
+            return stream.bytes.sha256Checksum
+        }
+    }
+}
+
+extension ManifestLoader {
+    struct EvaluationResult: Codable {
+        /// The path to the diagnostics file (.dia).
+        ///
+        /// This is only present if serialized diagnostics are enabled.
+        var diagnosticFile: AbsolutePath?
+
+        /// The output from compiler, if any.
+        ///
+        /// This would contain the errors and warnings produced when loading the manifest file.
+        var compilerOutput: String?
+
+        /// The manifest in JSON format.
+        var manifestJSON: String?
+
+        /// Any non-compiler error that might have occurred during manifest loading.
+        ///
+        /// For e.g., we could have failed to spawn the process or create temporary file.
+        var errorOutput: String? {
+            didSet {
+                assert(self.manifestJSON == nil)
+            }
+        }
+
+        var hasErrors: Bool {
+            return self.manifestJSON == nil
+        }
+    }
+}
+
+extension ManifestLoader {
+    /// Represents behavior that can be deferred until a more appropriate time.
+    struct DelayableAction<T> {
+        var target: T?
+        var action: ((T) -> Void)?
+
+        func perform() {
+            if let value = target, let cleanup = action {
+                cleanup(value)
+            }
+        }
+
+        mutating func delay() -> DelayableAction {
+            let next = DelayableAction(target: target, action: action)
+            target = nil
+            action = nil
+            return next
+        }
+    }
+}
+
```

Comment:
this data structs moved from line the code to the bottom to make code more readable. no function changes in them